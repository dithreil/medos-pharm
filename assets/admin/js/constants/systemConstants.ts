import {IPagination} from "../interfaces/request-params";

const selectPagination: IPagination = {
    descending: false,
    limit: 10,
    page: 1,
    rowsNumber: 10,
    rowsPerPage: null,
    sortBy: null,
};
export default {
    selectPagination,
}
