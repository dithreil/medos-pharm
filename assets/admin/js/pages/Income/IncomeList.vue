<template>
  <div class="page-bottom-margin">
    <div class="row no-wrap justify-between items-center">
      <h1 class="text-h4 text-primary">
        Список поступлений
      </h1>
      <q-btn
          flat
          color="primary"
          label="Создать поступление товаров"
          :to="{name: 'IncomeCreate'}"
      />
    </div>
    <q-table
        :data="tableData"
        :columns="tableColumns"
        :pagination.sync="pagination"
        :loading="loading"
        :filter="search"
        no-results-label="Список поступлений пуст"
        binary-state-sort
        @request="fetchIncomes"
    >
      <template #top-right>
        <q-input
            v-model="search"
            dense
            debounce="300"
            placeholder="Поиск"
            style="min-width: 300px;"
        >
          <template #prepend>
            <q-icon name="search"/>
          </template>
        </q-input>
      </template>
      <template #body-cell-actions="props">
        <q-td :props="props">
          <div class="row items-start q-gutter-xs">
            <q-btn
                dense
                flat
                color="primary"
                icon="edit"
                title="Редактировать потсупление"
                :to="{name: 'IncomeEdit', params: {incomeId: props.row.id}}"
            />
            <q-btn
                dense
                flat
                color="grey"
                title="Подробнее"
                icon="info"
                :to="{name: 'IncomeDetails', params: {incomeId: props.row.id}}"
            />
          </div>
        </q-td>
      </template>
    </q-table>
  </div>
</template>

<script lang="ts">
import {mapActions, mapGetters} from 'vuex';
import {Component, Vue} from 'vue-property-decorator';
import {IPagination, IRequestParams} from '../../interfaces/request-params';
import {IDocumentIncome, IDocumentIncomeData} from '../../interfaces/income';


@Component({
    computed: {
        ...mapGetters({
            incomesData: 'income/incomesData',
            incomeRequestParams: 'income/incomeRequestParams',
        }),
    },
    methods: {
        ...mapActions({
            updateIncomeRequestParams: 'income/updateIncomeRequestParams',
        }),
    },
})
export default class IncomeList extends Vue {
  protected incomesData!: IDocumentIncomeData;
  protected updateIncomeRequestParams!: (payload: IRequestParams) => any;

  private tableColumns = [
      {name: 'name', align: 'left', label: 'Дата поступления', field: 'date', sortable: true},
      {name: 'supplier', align: 'left', label: 'Поставщик', field: (row: IDocumentIncome) => row.supplier.name},
      {name: 'store', align: 'left', label: 'Торговая точка', field: (row: IDocumentIncome) => row.store.name},
      {name: 'actions', align: 'left', label: 'Действия'},
  ];
  private search = '';
  private pagination: IPagination = {
      sortBy: 'date',
      descending: false,
      page: 1,
      rowsPerPage: 10,
      rowsNumber: 0,
      limit: null,
  };
  private loading = false;

  get tableData() {
      return this.incomesData?.items || [];
  };

  mounted() {
      this.fetchIncomes({pagination: this.pagination});
  };

  async fetchIncomes({pagination}: { pagination: IPagination }) {
      this.loading = true;

      pagination.limit = pagination.rowsPerPage;
      pagination.rowsPerPage = null;

      await this.updateIncomeRequestParams({
          ...pagination,
          filter: this.search,
      });
      this.pagination.rowsNumber = this.incomesData.total;
      this.pagination.rowsPerPage = this.incomesData.limit;
      this.pagination.page = this.incomesData.page;
      this.pagination.sortBy = pagination.sortBy;
      this.pagination.descending = pagination.descending;

      this.loading = false;
  };
};
</script>
