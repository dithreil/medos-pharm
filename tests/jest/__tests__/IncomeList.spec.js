import {mountQuasar} from '@quasar/quasar-app-extension-testing-unit-jest';
import IncomeList from '../../../assets/admin/js/pages/Income/IncomeList.vue';
import Vuex from 'vuex';
import incomeData from '../test-data/incomeData';
import {createLocalVue} from '@vue/test-utils';
import {QBtn, QTable} from 'quasar';
import systemConstants from '../../../assets/admin/js/constants/systemConstants';
const localVue = createLocalVue();
localVue.use(Vuex);

describe('IncomeList', () => {
    let wrapper;
    let store;
    let actions;
    function mountComponent(store) {
        wrapper = mountQuasar(IncomeList, {
            quasar: {
                components: {
                    QTable, QBtn,
                },
            },
            mount: {
                store,
                localVue,
            },
        });
    }
    beforeEach(async()=> {
        actions = {
            'income/updateIncomeRequestParams': jest.fn(),
        };
        store = new Vuex.Store({
            actions,
            getters: {
                'income/incomesData': () => incomeData,
                'income/incomeRequestParams': () => null,
            },
        });
        mountComponent(store);
    });

    test('matches snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    test('tableData correct work ', async() => {
        expect(wrapper.vm.tableData).toBe(incomeData.items);

        store.getters = {
            'income/incomesData': () => null,
            'income/incomeRequestParams': () => null,
        };
        mountComponent(store);

        expect(wrapper.vm.tableData).toEqual([]);
    });

    test('fetchIncomes correct work ', async() => {
        expect(actions['income/updateIncomeRequestParams']).toBeCalledTimes(1);

        await wrapper.vm.fetchIncomes({pagination: systemConstants.selectPagination});


        expect(actions['income/updateIncomeRequestParams']).toBeCalledTimes(2);

        await wrapper.vm.fetchIncomes({pagination: systemConstants.selectPagination});

        expect(wrapper.vm.$data.pagination).toEqual({
            sortBy: null,
            descending: false,
            page: 1,
            rowsPerPage: 10,
            rowsNumber: 6,
            limit: 10,
        });
    });
});
