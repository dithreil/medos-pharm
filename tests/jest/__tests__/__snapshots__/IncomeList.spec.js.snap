// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeList matches snapshot 1`] = `
<div
  class="page-bottom-margin"
>
  <div
    class="row no-wrap justify-between items-center"
  >
    <h1
      class="text-h4 text-primary"
    >
      
      Список поступлений
    
    </h1>
     
    <q-btn-stub
      align="center"
      color="primary"
      flat="true"
      label="Создать поступление товаров"
      ripple="true"
      to="[object Object]"
    />
  </div>
   
  <q-table-stub
    binarystatesort="true"
    color="grey-8"
    columns="[object Object],[object Object],[object Object],[object Object]"
    data="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    filter=""
    filtermethod="[Function]"
    noresultslabel="Список поступлений пуст"
    pagination="[object Object]"
    rowkey="id"
    rowsperpageoptions="5,7,10,15,20,25,50,0"
    selected=""
    selection="none"
    separator="horizontal"
    sortmethod="[Function]"
  />
</div>
`;
